File: BaekJoon/Java/1182_부분수열의합_BackTracking.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

// 부분수열의합 1182 BackTracking
// O( 2^N )


public class Main {
    static int n;
    static int s;
    static int cnt = 0;
    static int[] arr;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        n = Integer.parseInt(st.nextToken());
        s = Integer.parseInt(st.nextToken());

        arr = new int[n];
        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < n; ++i) {
            arr[i] = Integer.parseInt(st.nextToken());
        }

        BackTracking(0, 0);

        if (s == 0) --cnt;

        System.out.println(cnt);
    }

    public static void BackTracking(int idx, int curSum) {
        if (idx == n) {
            if (curSum == s) ++cnt;
            return;
        }

        BackTracking(idx + 1, curSum);
        BackTracking(idx + 1, curSum + arr[idx]);
    }
}

File: BaekJoon/1182_부분수열의합_BackTracking.cpp
#include <iostream>
#include <vector>

using namespace std;

// 부분수열의합 1182 BackTracking
// O( 2^N )


void BackTracking(int idx, int curSum, int& cnt, 
    const int& n, const int& s, const vector<int>& v)
{
    // 모든 요소를 검사했을 때
    if (idx == n)
    {
        // 합이 S 와 같은 부분수열이 있다면 cnt 증가
        if (curSum == s) ++cnt;
        return;
    }

    // 현재 원소를 포함하지 않는 경우
    BackTracking(idx + 1, curSum, cnt, n, s, v);

    // 현재 원소를 포함하는 경우
    BackTracking(idx + 1, curSum + v[idx], cnt, n, s, v);
}

int main(void)
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);
    
    int n, s;
    cin >> n >> s;

    vector<int> v(n);
    for (int& e : v)
    {
        cin >> e;
    }

    int cnt{ 0 };

    BackTracking(0, 0, cnt, n, s, v);

    // s 가 0인 경우 공집합은 제외
    if (s == 0) --cnt;

    cout << cnt;


    return 0;
}
